src/bootstrap/tool.rs

```diff
@@ -227,8 +225,58 @@ impl Step for RemoteTestServer {
 }
 
 #[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct Rustdoc {
+    pub target_compiler: Compiler,
+}
+
+impl Step for Rustdoc {
+    type Output = PathBuf;
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun) -> ShouldRun {
+        run.path("src/tools/rustdoc")
+    }
+
+    fn make_run(run: RunConfig) {
+        run.builder.ensure(Rustdoc {
+            target_compiler: run.builder.compiler(run.builder.top_stage, run.host),
+        });
+    }
+
+    fn run(self, builder: &Builder) -> PathBuf {
+        let target_compiler = self.target_compiler;
+        let build_compiler = if target_compiler.stage == 0 {
+            target_compiler
+        } else {
+            builder.compiler(target_compiler.stage - 1, target_compiler.host)
```

Comment:
Sure! I'll add a comment as well once you're satisfied with my explanation. `rustdoc` is basically just a fancy `rustc`, so what we do here is mostly the same as in the `compile::Assemble` step. When building rustdoc for a given `stageN/bin` we want to do so with the compiler  for `stage(N-1)/bin` so we subtract 1 here. Before, this was done implicitly since rustdoc was built alongside the `rustc` by that `stage(N-1)` compiler.